midx.c

```diff
@@ -1110,7 +1111,113 @@ int expire_midx_packs(const char *object_dir)
 	return result;
 }
 
-int midx_repack(const char *object_dir, size_t batch_size)
+struct time_and_id {
+	timestamp_t mtime;
+	uint32_t pack_int_id;
+};
+
+static int compare_by_mtime(const void *a_, const void *b_)
 {
+	const struct time_and_id *a, *b;
+
+	a = (const struct time_and_id *)a_;
+	b = (const struct time_and_id *)b_;
+
+	if (a->mtime < b->mtime)
+		return -1;
+	if (a->mtime > b->mtime)
+		return 1;
 	return 0;
 }
+
+int midx_repack(const char *object_dir, size_t batch_size)
+{
+	int result = 0;
+	uint32_t i, packs_to_repack;
+	size_t total_size;
+	struct time_and_id *pack_ti;
+	unsigned char *include_pack;
+	struct child_process cmd = CHILD_PROCESS_INIT;
+	struct strbuf base_name = STRBUF_INIT;
+	struct multi_pack_index *m = load_multi_pack_index(object_dir, 1);
```

Comment:
I can't tell from my view of the code, so I'll just ask.

Where/How does the existing MIDX file on disk get locked and held between our load() and the write() at the bottom?  I see inside a lockfile inside write_midx_internal(), but that is only used during the parts of write().
I would think we'd need a way to disable/block other consumers from starting another read+write operation
concurrently.  I'm thinking specifically of 2 concurrent midx-repack commands racing.

Reply:
True, but what stops two `git repack` commands from racing?

We handle this in the VFS for Git side using a file-based lock.

Reply:
I guess I was more thinking about upstream which doesn't have a global lock,
than VFS4G which does.

Reply:
Right. Upstream doesn't have a lock, except for specific files being written. They rely on the user to not run some things concurrently, like `repack`.